### ISSUE TITLE
Increasing mem and no output files

### ISSUE THREAD

**Issue by pbatk** [rom1504/img2dataset#49]
Currently using your tool to download laion dataset, thank you for your contribution. The program grows in memory until it uses all of my 32G of RAM and 64G of SWAP. No tar files are ever output. Am I doing something wrong?

Using the following command (slightly modified from official command provided by laion)
`img2dataset --url_list laion400m-meta --input_format "parquet" \ 
              --url_col "URL" --caption_col "TEXT" --output_format webdataset \
                --output_folder webdataset --processes_count 1 --thread_count 12 --image_size 384 \
                  --save_additional_columns '["NSFW","similarity","LICENSE"]'`

**Comment by pbatk** [comment]
To be entirely fair, I have about 8G of RAM in use already so  more link 24G of RAM and 64G of SWAP

**Comment by rom1504** [comment]
> --processes_count 1 --thread_count 12

this is very low, you should use one process per core, and probably 128 threads

You might also want to add --enable_wandb to make it easier to check the performance

about the memory usage, it's normal that it would use some GB, but not that much. How is the cpu usage and bandwidth usage ?

**Comment by pbatk** [comment]
I actually dropped those super low to try and reduce the mem usage. I had it at 16 processes (core count) and 128 threads as suggested but it mem crashed in under 10 minutes

CPU stays around 25% usage and bandwidth cycles at 1Gb/s for about 20 seconds and then drops to 100Mb/s for 3-5 minutes

**Comment by pbatk** [comment]
Thank you for the rapid response and help!!!

**Comment by rom1504** [comment]
What do you see in the console as output ? (if you enable wandb and rerun, you could share the link here, that might be easier)

**Comment by pbatk** [comment]
Console output says : `Downloading file number 1 of 32 called laion400m-meta/part-00000-5b54c5d5-bbcf-484d-a2ce-0d6f73df1a36-c000.snappy.parquet`
It crashes before starting another file

Is there a way to DM you? I'd prefer not to share my name which is within my wandb.

**Comment by rom1504** [comment]
I guess on discord yes.
But anyway the issue seems to be coming before any downloading too place if that's the only thing that is printed
Here is an example of run that I've just started now to check the latest code https://wandb.ai/rom1504/img2dataset/runs/225ckjje?workspace=user-

It takes about 15GB of ram with these settings on my side.
What version of python are you using ?

**Comment by rom1504** [comment]
you can send the wandb link on my email I guess (it's on my github profile)

**Comment by rom1504** [comment]
I just checked with the latest metadata files and latest code and indeed I'm observing the same (too much memory usage at initialization)
That's not normal, I'll look into it and fix it

**Comment by pbatk** [comment]
I am using python 3.7. Thanks for helping, I'm sending the link now

**Comment by Skylion007** [comment]
@rom1504 Do you think it's due to Python not properly cleaning subprocs? 5 tasks_per_child could be too low if that's the case.

**Comment by pbatk** [comment]
Unfortunately I cannot figure out how to share the wandb with you yet. I will continue to work on it.

Additional info, I was using 1.9.1 & 1.9.2. After downgrading to 1.7.0 the problem resolved

**Comment by rom1504** [comment]
I found a silly bug due to a recent commit affecting downloading with multiple files. Will release a fix in a few minutes
5 tasks_per_child is ok because 1 shard takes about 1min at best

**Comment by rom1504** [comment]
it's now fixed and released as 1.9.5
@pbatk can you update and try again to confirm the fix ?

**Comment by pbatk** [comment]
I already launched with 1.7.0...
I can try on my laptop if that is meaningful

**Comment by rom1504** [comment]
I'd advise to restart with 1.9.5 as there are been a few improvements that will be useful to you

**Comment by rom1504** [comment]
the issue seems to be fixed to me

do say if you have more issues, thanks for reporting this one!

good luck with laion400m, I hope you can do interesting things with it!

**Comment by pbatk** [comment]
I'm restarting on your reccomendation. Will update on performance

**Comment by pbatk** [comment]
Update works great for me! One thought (this may be better for a separate issue), wouldn't you want to disable SSLverification? Or at least permit the user to disable it.

**Comment by rom1504** [comment]
that sounds like an interesting improvement. Would be useful to try it and measure the speed with/without. Do you think ssl verification impacts performance a lot ?
The current bottleneck is resizing (if you disable resizing then it's bandwidth)

**Comment by pbatk** [comment]
Not speed but throwaways for invalid SSL. So far I'm seeing about a 1% failure because of expired or invalid SSL. It might not be the most secure option so maybe it warrants a clear warning but still functionality that could be useful

**Comment by rom1504** [comment]
I see, yeah I'm interested to add it under an option (maybe disabled by default), if you want to PR it, feel free

**Comment by pbatk** [comment]
Ya, I can give it a go. I can definitely do it by tomorrow night
